doc: update documentation with default login details for MetalK8s UI & Grafana service (#2377)
scality/metalk8s: docs/operation/cluster_and_service_configuration.rst

To add a new static user, perform the following operations:

#. Generate a bcrypt hash of your new password.

   - Startup the MetalK8s utility container by refering to
     :ref:`this procedure <Troubleshooting-container>`

   - To generate the bcrypt hash, run the following command while replacing
     the <password> field with your plaintext password.

   .. code-block:: shell

>       root@bootstrap $ crictl exec -it \

Reviewer: Nah. The procedure to start the `metalk8s-utils` container (which I'll comment on separately) doesn't fix the Pod/container to a particular node, so `crictl` may not even work.

Also, it's not great form to pass passwords as CLI arguments, they easily end up in shell history and whatnot.

So, instead, why not

```
kubectl run htpasswd-$$ -i --image THE_IMAGE_WE_WANT --restart=Never --rm --command -- sh -c "htpasswd -nBC 12 '' | tr -d ':'"
```

Reviewer: Thinking of it, that won't work with taints.

Now, why don't we install `httpd-utils` on the bootstrap node, and document this package is required (in case someone wants to run this on another node)?

Author: Ok, i will add the package to the bootstrap node and be done

Author: Done

To add a new static user, perform the following operations:

   .. _Generate-password-hash:

#. Generate a bcrypt hash of your new password.

>    - To generate the bcrypt hash, On the Bootstrap node, run the following

Reviewer 2: ```suggestion
   - To generate the bcrypt hash, on the Bootstrap node, run the following
```

     command while replacing the <password> field with your plaintext password.

   .. code-block:: shell

>       root@bootstrap $ htpasswd -bnBC 14 "" <password> | tr -d ':'

Reviewer 2: I thought we said the best thing is to never input passwords in clear text on the CLI... So, better use the suggestion from Nicolas:

```suggestion
      root@bootstrap $ htpasswd -nBC 14 "" | tr -d ':'
      New password:
      Re-type new password:
      <your hash here, starting with "$2y$14$">
```

To add a new static user, perform the following operations:

#. Generate a bcrypt hash of your new password.

   .. code-block:: shell

>       root@bootstrap $ htpasswd -bnBC 12 "" <replace-with-password> | tr -d ':'

Reviewer 2: I think the point of this ticket was to not rely on some external package, but build this hash with embedded dependencies...

I'd rather ship a simple Salt module to generate it from the available tooling and document this (in the future, we'll be able to reuse such a module in our CLI).

Author: In regards to this, out of the box, Salt alone will not work here since the hash is a  `bcrypt` with work factor of atleast 12. Salt does not come with the bcrypt package as checked meaning we need to install this external package.

For a start, we could package bcrypt in the ISO, but we already have httpd-tools added to metalk8s-utils image as from this PR yesterday: https://github.com/scality/metalk8s/pull/2374 which seems sufficient for me given that we have docs on how to run the container

Btw, this ticket is really about updating the documentation to include default login credentials

Reviewer 2: Oh, right. You're changing the lines on how to provision users, so I thought you were also tackling #2352. If this PR really is only about default creds, then there's a lot of changes that are not related and would need another PR, don't you think?

Reviewer 2: If we want to use the `metalk8s-utils` container image for generating this, we could just provide the full command to run. Maybe (not tested yet) something like:

```shellscript
$ kubectl --kubeconfig=/etc/kubernetes/admin.conf \
    run -it tmp-gen-pwd \
    --image=metalk8s-registry-from-config.invalid/metalk8s-|release|/metalk8s-utils:|release| \
    --command -- htpasswd -bnBC 12 "" <replace-with-password> | tr -d ':'
```

Author: I probably will rewrite this PR to include your comments here: https://github.com/scality/metalk8s/pull/2377#discussion_r404694946 then close the two tickets

                      userID: "<uuidv4>"
      [...]

#. Save the ConfigMap changes.

#. From the Bootstrap node, run the following to propagate the
   changes.

   .. parsed-literal::

      root@bootstrap $ kubectl exec -n kube-system -c salt-master \\
                       --kubeconfig /etc/kubernetes/admin.conf \\

>                        salt-master-bootstrap -- bash -c  \\

Reviewer 2: Why the `bash -c` here? We shouldn't need it and be able to just run `salt-run ...` directly.

> #. Save and apply the changes.

Reviewer 3: isn't it save to apply the change? I mean by saving, the change will be applied right? otherwise, you would need to explain how to apply the change

Author: Ok, will keep it simple => `Save the changes`

Author: Done

#. Save and apply the changes.

#. From the Bootstrap node, run the following to propagate the
   changes.

   .. parsed-literal::

>       root@bootstrap $ crictl exec -it \\

Reviewer: Always prefer `kubectl ...` over `crictl`.

Author: fixed!

#. Verify that the user has been successfully added and you can log in to the
   MetalK8s UI using the new email and password.

.. _Change-dex-user-password:

Change Password for local static user
"""""""""""""""""""""""""""""""""""""

To change the password of an existing user, perform the following operations:

#. Generate a bcrypt hash of the new password.

   .. code-block:: shell

      root@bootstrap $ htpasswd -bnBC 14 "" <replace-with-password> | tr -d ':'

#. From the Bootstrap node, edit the ConfigMap ``metalk8s-dex-config`` and then

>    change the ``hash`` for the selected user:

Reviewer 2: Maybe show an example of the diff on a changed ConfigMap, to make it easier for users?

   .. code-block:: shell

      root@bootstrap $ kubectl --kubeconfig /etc/kubernetes/admin.conf \
                         edit configmaps metalk8s-dex-config -n metalk8s-auth

#. Save and apply the ConfigMap changes.

> #. From the Bootstrap node, execute the following command which connects to
>    the Salt master container and applies salt-states to propagate the new
>    changes down to the underlying services.

Reviewer 2: Let's keep it simple:

```suggestion
#. From the Bootstrap node, run the following to propagate the
   changes:
```

>    Make sure to replace the <version> prefix in the command below with the
>    currently installed MetalK8s version number e.g 2.5.0
> 
>    .. code-block:: shell
> 
>       root@bootstrap $ crictl exec -it \
>                          $(crictl ps -q --label io.kubernetes.container.name=salt-master) bash \
>                          -c "salt-run state.sls metalk8s.addons.dex.deployed saltenv=metalk8s-<version>"

Reviewer 2: Just use the current version in docs:

```suggestion
   .. parsed-literal::

      root@bootstrap $ crictl exec -it \
                         $(crictl ps -q --label io.kubernetes.container.name=salt-master) bash \
                         -c "salt-run state.sls metalk8s.addons.dex.deployed saltenv=metalk8s-|release|"
```

#. Verify that the password has been changed by accessing the related service

>    using the new password.

Reviewer 2: "the related service" is unclear. Let's simply suggest logging in the UI.

#. Verify that the user has been successfully added and you can log in to the
   MetalK8s UI using the new email and password.

.. _Change-dex-user-password:

Change Password for local static user
"""""""""""""""""""""""""""""""""""""

To change the password of an existing user, perform the following operations:

#. Generate a bcrypt hash of the new password.

   - Startup the MetalK8s utility container by refering to
     :ref:`this procedure <Troubleshooting-container>`

   - To generate the bcrypt hash, run the following command while replacing
     the <password> field with your plaintext password.

   .. code-block:: shell

>       root@bootstrap $ crictl exec -it \

Reviewer: See above. Also, de-duplicate this section with the one above. The way to calculate a password hash is always the same.

Author: Done!

                       $(crictl ps -q --label io.kubernetes.container.name=metalk8s-utils) bash \
                       -c "htpasswd -bnBC 12 "" <password> | tr -d ':'"

#. From the Bootstrap node, edit the ConfigMap ``metalk8s-dex-config`` and then
   change the ``hash`` for the selected user:

   .. code-block:: shell

      root@bootstrap $ kubectl --kubeconfig /etc/kubernetes/admin.conf \
                         edit configmaps metalk8s-dex-config -n metalk8s-auth

      [..]
      config.yaml: |-
         localuserstore:
            enabled: true
            userlist:
               - email: "admin@metalk8s.invalid"
                  hash: "<new-password-hash>"
                  username: "admin"
                  userID: "08a8684b-db88-4b73-90a9-3cd1661f5466"
      [...]

#. Save and apply the ConfigMap changes.

#. From the Bootstrap node, run the following to propagate the
   changes.

   .. parsed-literal::

>       root@bootstrap $ crictl exec -it \\

Reviewer: `kubectl`

   the Salt master container and applies salt-states to propagate the new
   changes down to the underlying services.

   .. note::

      Scaling the number of pods for services like Prometheus and Alertmanager
      require provisioning extra persistent volumes for these pods to startup
      normally. Refer to :ref:`this procedure <Provision Prometheus storage>`
      for more information.

   .. parsed-literal::

>       root@bootstrap $ crictl exec -it \\

Reviewer: `kubectl`
